kolena/_experimental/dataset/_dataset.py

```diff
@@ -83,6 +90,35 @@ def _infer_datatype_value(x: str) -> str:
     return DatapointType.TABULAR.value
 
 
+def _add_datatype(df: pd.DataFrame) -> None:
+    """Adds `data_type` column(s) to input DataFrame."""
+    prefixes = {
+        column.rsplit(sep=SEP, maxsplit=1)[0]
+        for column in df.columns.values
+        if isinstance(column, str) and SEP in column
+    }
+    if prefixes:
+        df[DATA_TYPE_FIELD] = DatapointType.COMPOSITE.value
+        for prefix in prefixes:
+            if prefix in df.columns:
+                raise InputValidationError(
+                    f"Conflicting column '{prefix}' encountered when formatting composite dataset.",
+                )
+            if SEP in prefix:
+                raise InputValidationError(
+                    f"More than one delimeter '{SEP}' in prefix: '{prefix}'.",
+                )
+
+            composite_columns = df.filter(regex=rf"^{prefix}", axis=1).columns.to_list()
+            composite = df.loc[:, composite_columns].rename(columns=lambda col: col.split(SEP)[-1])
+            composite[DATA_TYPE_FIELD] = _infer_datatype(composite)
+
+            df[prefix] = composite.to_dict("records")
+            df.drop(columns=composite_columns, inplace=True)
```

Comment:
Can we avoid using `inplace=True`. My general reading is that there is no clear cut performance benefit and with our usage of passing dataframe as input, it's easy to introduce hard-to-debug bugs because of inplace manipulation. And there's no way to tell function caller that we do that to their input.

Reply:
Although weary of the usage of `inplace=True` and prefer not to use it, this comment is nonblocking, as currently `_add_datatype` has single invocation point.

Reply:
`_add_datatype` is going to have a few more invocations on the backend so I think it is worthwhile to have the discussion now.  We can update the function to return a new copy of the DataFrame leaving the previous one intact, but if all functions along the dataset deser call stack ensure the input dataframe is not manipulated there will certainly be performance implications for large datasets.

In my opinion, the public api `register_dataset` should create a copy immediately and called private functions are free to mutate. This ensures that the users DataFrame is not mutated and keeps copies to a minimum.

Reply:
Even for private functions, they could be reused/shared by other public/private methods, now or future. A function that modifies input data is easy to cause bugs when call paths are non-trivial. Current scoping to `_add_datatype` (as small as possible) looks ok to me, still future developers should be aware of this behavior if they want to use it. I don't think we should expand it to `register_dataset`.

A slightly different discussion of use of `inplace=True`, I don't want to blindly believe in performance gain from `inplace`, but also don't want to unnecessarily fear of it. For dicussion's sake, references that in general not in favor of `inplace`
1. https://stackoverflow.com/questions/45570984/in-pandas-is-inplace-true-considered-harmful-or-not
2. https://sourcery.ai/blog/pandas-inplace/